keps/sig-api-machinery/0006-apply.md

```diff
@@ -194,8 +195,41 @@ atomic. That can be specified with `// +mapType=atomic` or `//
 +structType=atomic` respectively. They map to the same openapi extension:
 `"x-kubernetes-map-type": "atomic"`.
 
-#### Kubectl
-##### Server-side Apply
+#### Clients
+
+##### Apply Status Header
+
+Client side apply in kubectl determines if the applied object would result in any changes on the apiserver and provides the user with this information.
+
+To provided similar information with server-side apply, a dedicated header is set by the apiserver
+if a **successful apply request** caused changes to the existing resource.
+
+```txt
+X-Kubernetes-Apply-Success-Type: "changes-made" || "changes-unneeded"
+```
+
+The value may be,
+
+- `changes-unneeded` in cases where the request did not cause a change to the resource's managedFields.
```

Comment:
I would very clearly rephrase this, and consider how we're going to detect changes.

After giving it some thoughts, it seems clearly superior to me to compare the resourceVersion before and after going to etcd. That is the best and most efficient way to know if something actually changed.  (I'm also assuming that a no-op apply won't write to etcd, which I think we have tests for since it's a necessary requirement).

We can take that thought and go further: Apply is not the only verb where the request is made without knowing the current resourceVersion (hence client can't check that it changed):
- Create is an obvious case that is no worth discussing about,
- Update may (hopefully) or may not specify the resource version, I think it might be useful in the latter case to report if the resource has changed.
- All forms of PATCH, including apply, may or may not know the resourceVersion of the resource they are editing. In all these cases, it might be useful to return that information. Also note that this would literally fix **once and for all** issue https://github.com/kubernetes/kubernetes/issues/66450, since right now we decide if the resource is changed when we send a patch, even if the patch does nothing.

I don't know if "needed" is the right word then. `changes-made`?

Reply:
@lavalamp could you take a look at this comment?
I'm in favor of the points @apelisse made and would change the KEP based on them.
This also means including non-apply and therefor also look for a more generic header name.

I'm not sure if this change would make a replacement for what kubectl currently does. @apelisse mentioned the diff is made before applying. If kubectl would skip an apply if there is no expected change, that would be different.

Reply:
> All forms of PATCH, including apply, may or may not know the resourceVersion of the resource they are editing

The *user* might not know, but that doesn't seem important since *apiserver* definitely knows?

Reply:
Yeah, the question is, do we want the apiserver to let the user know? User feedback seems to confirm that they do want to know.